text/0000-cargo-features2.md

```diff
@@ -0,0 +1,514 @@
+a dependency appears multiple times in the dependency graph. The new behaviors
+are described below.
+
+For [target dependencies](#target-dependencies) and
+[dev-dependencies](#dev-dependencies), the general rule is, if a dependency is
+not built, it does not affect feature resolution. For [host
+dependencies](#host-dependencies), the general rule is that packages used for
+building (like proc-macros) do not affect the packages being built.
+
+The following three sections describe the new behavior for three difference
+situations.
+
+### Target dependencies
+
+When a package appears multiple times in the build graph, and one of those
+instances is a target-specific dependency, then the features of the
+target-specific dependency are only enabled if the target is currently being
+built. For example:
+
+```toml
+[dependency.common]
+version = "1.0"
+features = ["f1"]
+
+[target.'cfg(windows)'.dependencies.common]
+version = "1.0"
+features = ["f2"]
+```
+
+When building this example for a non-Windows platform, the `f2` feature will
+*not* be enabled.
+
+### dev-dependencies
+
+When a package is shared as a normal dependency and a dev-dependency, the
+dev-dependency features are only enabled if the current build is including
+dev-dependencies. For example:
+
+```toml
+[dependencies]
+serde = {version = "1.0", default-features = false}
+
+[dev-dependencies]
+serde = {version = "1.0", features = ["std"]}
+```
+
+In this situation, a normal `cargo build` will build `serde` without any
+features. When built with `cargo test`, Cargo will build `serde` with its
+default features plus the "std" feature.
+
+Note that this is a global decision. So a command like `cargo build
+--all-targets` will include examples and tests, and thus features from
+dev-dependencies will be enabled.
+
+### Host dependencies
+
+When a package is shared as a normal dependency and a build-dependency or
+proc-macro, the features for the normal dependency are kept independent of the
+build-dependency or proc-macro. For example:
+
+```toml
+[dependencies]
+log = "0.4"
+
+[build-dependencies]
+log = {version = "0.4", features=['std']}
+```
+
+In this situation, the `log` package will be built with the default features
+for the normal dependencies. As a build-dependency, it will have the `std`
+feature enabled. This means that `log` will be built twice, once without `std`
+and once with `std`.
+
+Note that a dependency shared between a build-dependency and proc-macro are
+still unified. This is intended to help reduce build times, and is expected to
+be unlikely to cause problems that feature unification usually cause because
+they are both being built for the host compiler, and are only used at build
+time.
+
+## Resolver opt-in
+
+Testing has been performed on various projects that shows sometimes
+dependencies are written to assume that features from another part of the
+graph are enabled, and fail to compile with the new resolver. Because the new
```

Comment:
It took me a few tries to parse this sentence. Maybe it can be broken up something like "Testing has been performed on various projects. Some whur found to fail to compile with the new resolver. Witch shows that sometimes dependencies are written to assume that features from another part of the graph are enabled."